### DISCUSSION TITLE
x-collapse not work correctly

### DISCUSSION THREAD

**Discussion by Sulaymon-dev** [alpinejs/alpine#2409] [7. Bugs (archived)]
Hello everyone. 
I use Alpine JS 3.5.1 and x-collapse plugin. When toggle content some times it's not work correctly:

```HTML
<div x-data="{isShow:true}" class="p-6">
	<button @click="isShow =!isShow">Toggle</button>
        <ul class="space-y-4 border-2" x-show="isShow" x-collapse>
        	<template x-for="i in 5">
                	<li x-text="'text' + i"></li>
                </template>
        </ul>
</div>
```

https://user-images.githubusercontent.com/40419449/143424229-35c5d9d0-a2dc-4922-81de-2b097eda4500.mp4

**Comment by dfallman** [comment]
We've been having various issues too with x-collapse and Tailwind up to the point where it's often been unusable unfortunately. However, a few things I've learnt so far:

- x-collapse seems to work best on a `div` where it's just it and the `x-show` and nothing else -- in your code I'd try renaming the `ul` to a `div`, insert the `ul` under it with all the tailwind stuff 
- make sure the page doesn't have any missing `</div>`s and similar things -- most browsers will try to render the page the best they can and glitchy behavior like this can (but doesn't have to) be the result of the browser struggling to render it 
- finally, if you're having issues and there are `<template>` tags around, always assume they're causing it :) -- this glitch might be triggered by x-collapse and the `x-for` rendering at the same time

**Comment by Sulaymon-dev** [comment]
There is probably an error in the following code:
```JavaScript
out(before = () => {}, after = () => {}) {
                let full = el.getBoundingClientRect().height

                Alpine.transition(el, setFunction, {
                    during: transitionStyles,
                    start: { height: full+'px' },
                    end: { height: floor+'px' },
                }, () => {}, () => {
                    el._x_isShown = false

                    // check if element is fully collapsed
                    if (el.style.height == `${floor}px`) {
                        Alpine.nextTick(() => {
                            Alpine.setStyles(el, {
                                display: 'none',
                                overflow: 'hidden'
                            })
                            el.hidden = true;
                        })
                    }
                })
            },
```
So that after transition the element is displayed in full again and hide:

https://user-images.githubusercontent.com/40419449/143529363-66054318-b976-49b3-b17a-99aad9bd4813.mp4

**Comment by calebporzio** [comment]
Hi @Sulaymon-dev, I tried replicating with your code and I couldn't. Can you verify this is still an issue on version 3.5.2 of both Alpine and the plugin?

**Reply by Sulaymon-dev** [reply]
I tested this on the Core I5-8250U, Core I3-3110m, and Redmi 8 (Snapdragon 439):
https://codesandbox.io/s/nervous-bose-hqwsk?file=/index.html
In Core I5-8250U nearly 1 in 10 toggle did not work correctly, if you slowdown CPU (4x, 6x), this number will increase 4,7...

This bug did not exist in version 3.4.2 and below...

**Reply by SimoTod** [reply]
@Sulaymon-dev can you stress test this code please? https://codepen.io/SimoTod/pen/rNGNPYP
Thanks

**Reply by philc-d** [reply]
This issue exists on the x-collapse example in the docs too [https://alpinejs.dev/plugins/collapse#x-collapse](url) ("Toggle Content" a couple of times and flashing will occur)

**Reply by Sulaymon-dev** [reply]
@SimoTod I tested this on phone redmi 8. From 12 toggle 3 toggle not work correctly...

https://user-images.githubusercontent.com/40419449/144155799-1b2749af-f51b-493f-8495-8f2d179b68e8.mp4

**Reply by SimoTod** [reply]
Thanks @Sulaymon-dev 
I've updated https://codepen.io/SimoTod/pen/rNGNPYP and it should work as expe ted now. Can you try again, please?

**Reply by Sulaymon-dev** [reply]
It still has that bug. :-(

**Reply by Sulaymon-dev** [reply]
In this case it's executed correctly

```html
<div x-data="{isShow:true}" class="p-6">
  <button @click="isShow =!isShow">Toggle</button>
  <div :class="!isShow && 'overflow-hidden'">
    <ul class="space-y-4 border-2" x-show="isShow" x-collapse>
    <template x-for="i in 5">
      <li x-text="'text' + i"></li>
    </template>
  </ul>
  </div>
</div>
```

**Reply by SimoTod** [reply]
It's weird, it works correctly locally even with a busy CPU.

**Reply by SimoTod** [reply]
Oh, silly me. I had `<script src="https://unpkg.com/@alpinejs/collapse@3.x.x/dist/cdn.min.js">` in the codepen so it was always overriding it with the bugged version. 🤦‍♂️

**Reply by SimoTod** [reply]
Sorry about that, it should be updated for real now.

**Reply by Sulaymon-dev** [reply]
It's work great now 👍👍👍

**Reply by SimoTod** [reply]
Brilliant, I'll tidy it up and send a PR today
